### ISSUE TITLE
rake routes fails when route constraint regex contains comments

### ISSUE THREAD

**Issue by tommyp** [rails/rails#18402]
When the constraint regex on a route contains comments, `rake routes` fails with the following error:

```
RegexpError: end pattern with unmatched parenthesis: /^\/entry\/((?x-mi:[a-f\d]{8}-[a-f\d]{4}-[1-5]#Version:http:\/\/tools.ietf.org\/html\/rfc4122#section-4.1.3[a-f\d]{3}-[89ab]#Variant:http:\/\/tools.ietf.org\/html\/rfc4122#section-4.1.1[a-f\d]{3}-[a-f\d]{12}))$/
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/actionpack-4.1.5/lib/action_dispatch/routing/inspector.rb:52:in `initialize'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/actionpack-4.1.5/lib/action_dispatch/routing/inspector.rb:52:in `new'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/actionpack-4.1.5/lib/action_dispatch/routing/inspector.rb:52:in `json_regexp'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/actionpack-4.1.5/lib/action_dispatch/routing/inspector.rb:133:in `block in collect_routes'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/actionpack-4.1.5/lib/action_dispatch/routing/inspector.rb:126:in `collect'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/actionpack-4.1.5/lib/action_dispatch/routing/inspector.rb:126:in `collect_routes'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/actionpack-4.1.5/lib/action_dispatch/routing/inspector.rb:93:in `format'
	from (irb):4
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/railties-4.1.5/lib/rails/commands/console.rb:90:in `start'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/railties-4.1.5/lib/rails/commands/console.rb:9:in `start'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/railties-4.1.5/lib/rails/commands/commands_tasks.rb:69:in `console'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/railties-4.1.5/lib/rails/commands/commands_tasks.rb:40:in `run_command!'
	from /usr/lib/rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/railties-4.1.5/lib/rails/commands.rb:17:in `<top (required)>'
	from bin/rails:4:in `require'
	from bin/rails:4:in `<main>'
```

The code in question is [here](https://github.com/alphagov/content-register/blob/2ced4612654a0a18d97b881a6ee52684f76a0c2f/config/routes.rb#L4) and as follows:

```ruby
Rails.application.routes.draw do
  with_options :format => false do |r|
    r.get '/entries' => 'entries#index'
    r.put '/entry/:content_id' => 'entries#update', constraints: { content_id: Entry::UUID_REGEX }

    r.get '/healthcheck' => proc {|env| [200, {}, ["OK"]]}
  end
end
```

The regex is [here](https://github.com/alphagov/content-register/blob/2ced4612654a0a18d97b881a6ee52684f76a0c2f/app/models/entry.rb#L4) and as follows:

```ruby
UUID_REGEX = %r{
    [a-f\d]{8}
    -
    [a-f\d]{4}
    -
    [1-5]   # Version: http://tools.ietf.org/html/rfc4122#section-4.1.3
    [a-f\d]{3}
    -
    [89ab]  # Variant: http://tools.ietf.org/html/rfc4122#section-4.1.1
    [a-f\d]{3}
    -
    [a-f\d]{12}
  }x
```

The line that it fails on in the source is [here](https://github.com/rails/rails/blob/v4.1.5/actionpack/lib/action_dispatch/routing/inspector.rb#L52). I know this isn't the latest version of Rails but that method looks like it hasn't changed at all compared to [master](https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/routing/inspector.rb#L45).

**Comment by rohit** [comment]
A sample app to reproduce the error. https://github.com/rohit/sample-regex-routes-18402

IIRC the rake tasks don't have tests? I might be mistaken about this though.

**Comment by rohit** [comment]
In my explorations I found the following which might help:

1. Applying the `json_regexp` method to the original regexps i.e either `TEST_REGEX` or `UUID_REGEX` doesn't raise any errors.
2. The regexp handled by the method has the path added to it like `/users/`.
3. Removing the regexp option `x` doesn't raise the error.
4. A diff of the value of `str` obtained by applying the `json_regexp` method to the path infused regexp in the normal execution of rake task vs that obtained by applying to the original regexp in plain IRB:

```diff
-"^\\/users\\/((?x-mi:[a-f\\d]{8}-[a-f\\d]{4}-[1-5]#Version:http:\\/\\/tools.ietf.org\\/html\\/rfc4122#section-4.1.3[a-f\\d]{3}-[89ab]#Variant:http:\\/\\/tools.ietf.org\\/html\\/rfc4122#section-4.1.1[a-f\\d]{3}-[a-f\\d]{12}))(?:\\.([^\\/.?]+))?$"
+"[a-f\\d]{8}-[a-f\\d]{4}-[1-5]#Version:http:\\/\\/tools.ietf.org\\/html\\/rfc4122#section-4.1.3[a-f\\d]{3}-[89ab]#Variant:http:\\/\\/tools.ietf.org\\/html\\/rfc4122#section-4.1.1[a-f\\d]{3}-[a-f\\d]{12}"
```

**Comment by rohit** [comment]
I did some more exploring. Firstly I was wrong, there are tests for the regexp converter at https://github.com/rails/rails/blob/master/actionpack/test/dispatch/routing/inspector_test.rb#L29

So I wrote a simple test that raises the error:

```ruby
      def test_json_regexp_converter_constraint_with_comments_ignored
        constraint_regexp = %r{
          a #a
        }x
        @set.draw do
          get '/cart/:slug', :to => 'cart#show', :constraints => { :slug => constraint_regexp }
        end
        route = ActionDispatch::Routing::RouteWrapper.new(@set.routes.first)
        assert route.json_regexp
      end
```

Note the regexp is much simpler this time. This raises the same error as mentioned by the OP.

The problem causing substitution in `json_regexp` is:

```ruby
      def json_regexp
        str = regexp.inspect.
              sub('\\A' , '^').
              sub('\\Z' , '$').
              sub('\\z' , '$').
              sub(/^\// , '').
              sub(/\/[a-z]*$/ , '').
              gsub(/\(\?#.+\)/ , '').
              gsub(/\(\?-\w+:/ , '(').
              gsub(/\s/ , '') # <=== THIS ONE!
        Regexp.new(str).source
      end
```

By removing/commenting-out the substitution that removes the spaces the test does not raise an error. The difference between the output of `json_regexp` with and without the removal of spaces gsub call is:

```diff
- "^\\/cart\\/((?x-mi:\n          a #a\n        ))(?:\\.([^\\/.?]+))?$"
+ "^\\/cart\\/((?x-mi:a#a))(?:\\.([^\\/.?]+))?$"
```

I'm guessing it has something to do with the embedded regexp option `?x-mi:` although I have no idea.

Also worth noting that the regex with the `?x-mi:` option is not compatible with Javascript regexp engine (based on limited trial).

**Comment by brainopia** [comment]
@tommyp @rohit Hi, I've sent a PR with a fix #18434
